File: cpp/240_Searcha2DMatrixII.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if(matrix.empty() || matrix[0].empty()) return false;
        return binarySearchMatrix(matrix,0,0,matrix.size()-1,matrix[0].size()-1,target);
    }
private:
    bool binarySearchMatrix(vector<vector<int>>& matrix, int li, int lj, int hi, int hj, int target){
        if(li > hi || lj > hj) return false;
        int mi = (li + hi)/2;
        int mj = (lj + hj)/2;
        if(matrix[mi][mj] == target) return true;
        else if(matrix[mi][mj] > target){
            return binarySearchMatrix(matrix,li,lj,hi,mj-1,target) || binarySearchMatrix(matrix,li,mj,mi-1,hj,target);
        }else{
            return binarySearchMatrix(matrix,li,mj+1,hi,hj,target) || binarySearchMatrix(matrix,mi+1,lj,hi,mj,target);
        }
        return false;
    }
};

File: js/240_Searcha2DMatrixII.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if(matrix.length < 0 || matrix[0].length < 0) return false;
    return binarySearchMatrix(matrix,0,0,matrix.length-1,matrix[0].length-1,target);
};

var binarySearchMatrix = function(matrix,li,lj,hi,hj,target) {
    if(li > hi || lj > hj) return false;
    var mi = Math.floor((li + hi)/2), mj = Math.floor((lj + hj)/2);
    if(matrix[mi][mj] == target) return true;
    else if(matrix[mi][mj] > target){
        return binarySearchMatrix(matrix,li,lj,hi,mj-1,target) || binarySearchMatrix(matrix,li,mj,mi-1,hj,target);
    }else{
        return binarySearchMatrix(matrix,li,mj+1,hi,hj,target) || binarySearchMatrix(matrix,mi+1,lj,hi,mj,target);
    }
    return false;
}